server/aws-lsp-identity/src/language-server/ssoTokenAutoRefresher.ts

```diff
@@ -1,26 +1,65 @@
-import { SsoTokenChangedParams } from '@aws/language-server-runtimes/protocol'
-import { RefreshingSsoCache } from '../sso/cache/refreshingSsoCache'
+import { SsoSession, SsoTokenChangedParams } from '@aws/language-server-runtimes/protocol'
+import { RefreshingSsoCache, refreshWindowMillis, retryWindowMillis } from '../sso/cache/refreshingSsoCache'
+import { throwOnInvalidClientName, throwOnInvalidSsoSession, throwOnInvalidSsoSessionName } from '../sso/utils'
+
+const bufferedRefreshWindowMillis = refreshWindowMillis * 0.95
+const bufferedRetryWindowMillis = retryWindowMillis * 1.05
+const maxRefreshJitterMillis = 10000
+const maxRetryJitterMillis = 3000
 
 export type RaiseSsoTokenChanged = (params: SsoTokenChangedParams) => void
 
-export class SsoTokenAutoRefresher {
-    constructor(
-        private readonly cache: RefreshingSsoCache,
-        private readonly raiseSsoTokenChanged: RaiseSsoTokenChanged
-    ) {}
-
-    watch(ssoSessionName: string): void {
-        // https://nodejs.org/api/timers.html#scheduling-timers
-        // Try to load token from cache
-        // If couldn't load token throw E_INVALID_TOKEN
-        // If token is expired throw E_TOKEN_EXPIRED
-        // setTimeout for 1 hour - (4 - 5 minutes jittered) to attempt refresh via RefreshingSsoCache
-        // Store token and Timeout
+export class SsoTokenAutoRefresher implements Disposable {
+    private readonly timeouts: Record<string, NodeJS.Timeout> = {}
+
+    constructor(private readonly ssoCache: RefreshingSsoCache) {}
+
+    [Symbol.dispose](): void {
+        for (const ssoSessionName of Object.keys(this.timeouts)) {
+            this.unwatch(ssoSessionName)
+        }
+    }
+
+    async watch(clientName: string, ssoSession: SsoSession): Promise<void> {
+        throwOnInvalidClientName(clientName)
+        throwOnInvalidSsoSession(ssoSession)
+
+        this.unwatch(ssoSession.name)
+
+        const ssoToken = await this.ssoCache.getSsoToken(clientName, ssoSession).catch(_ => undefined)
+
+        // Token doesn't exist, was invalidated by another process, or has expired
+        if (!ssoToken) {
+            return
+        }
+
+        const nowMillis = Date.now()
+        const expiresAtMillis = Date.parse(ssoToken.expiresAt)
+        let delayMillis: number
+
+        if (nowMillis < expiresAtMillis - refreshWindowMillis) {
```

Comment:
1. Should we create a common utility for jitters?
2. `Math.random()` will ensure the jitter is between 0 to ~`1m66sec`. Is that enough?

Reply:
Jitter can be a bit more variable whether it occur +/- or just one one side and the timing is dependent on the application.  For now, I'm going to leave it in-line, but if I find it starts getting reused more often, I will refactor it into a more reusable form.